rfcs/0154-Migrate-queue-to-postgres.md

```diff
@@ -0,0 +1,102 @@
+* Azure is unreliable with frequent failures and inscrutible rate-limiting. While we are resilient
+to these errors, they do cause inefficiency and sometimes cascading failures
+* Storing data in postgres will enable features users desire:
+  * Queue introspection
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status
+    * What tasks has this worker performed?
+  * Ad-hoc analytics
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically changing task levels
+
+# Details
+
+## Transition
+
+The transition to Postgres will be done in two steps:
+1. Transfer the data from Azure to Postgres without data loss. This will not aim
+   at normalizing the data but rather prioritize a safe and easy transition.  In
+   other words, the data will be backed by Postgres, but still behaving like
+   Azure.
+2. Go fully into normalized SQL tables.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
+
+## Schema Migration
+
+Taskcluster will write its own tooling to handle schema migrations and will have
+all migrations tested with sample data to ensure we get the expected schema. In
+addition, there will be one schema for the whole monorepo, so that schema
+updates happen atomically across all services.
+
+## Backups and Restores
+
+Teams operating Taskcluster will rely on the cloud provider's backup system to
+handle backups and restores.
+
+## DB Version Upgrades Compatibility
+
+When deploying a new version of Taskcluster, any necessary database upgrades
+*must* be applied before any services are upgraded.  Taskcluster will maintain
+the invariant that services expecting database version S can interoperate with a
+database at version V as long as V >= S.
+
+This invariant is maintained through careful attention to the definitions of the
+stored procedures (user-defined functions). These stored procedures are changed
+as necessary during the upgrade process, such that a procedure operates
+identically before and after the ugprade: identical arguments, identical return
+type, and correct behavior. For example, if an upgrade factors a single table
+into two tables, then a procedure which previously queried from the single table
+would be rewritten during the ugprade to perform a join over the two new tables,
+returning appropriate results in the same format as before the upgrade.
```

Comment:
I think it's worth adding another example. Let's say you have a query that returns column foo, but in the next version your code wants to know the value of column bar as well. As I understand this process, you would introduce a new stored procedure to return both columns and stop using the existing stored procedure that returned the single column. That new stored procedure is then deployed before the code that uses it is deployed.

Reply:
Correct. I'll update the RFC to include your example.